keps/sig-storage/20200116-node-failover.md

```diff
@@ -0,0 +1,191 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.
+
+This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.
+
+## Motivation
+
+Today kubernetes doesn't enable providers that are able to safely detach (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Implement in-cluster logic to handle node/control plane partitioning 
+* enable detach for all the storage providers
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+	// SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations, when set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking for detach safety
+	// +optional
+  SafeDetach *bool
+}
+
+```
+
+we will also introduce a new field for `StatefulSet`:
+
+```go
+// A StatefulSetSpec is the specification of a StatefulSet.
+type StatefulSetSpec struct {
+  // tolerateDuplication enables relaxing the uniqueness guarentees
+  // of stateful pods. this enables trading uniqueness of the replicas
+  // against more availability when the node is partioned
+  tolerateDuplication *bool
+}
+```
+
+This field is needed to identify which workloads toleration duplication, since forceful deletion would trigger immediate deletion and re-creation by the statefulset controller.
```

Comment:
I don't recall discussing this part, sorry.  I don't know the StatefulSet API well enough to know whether this is the best representation of this idea or if it is even needed.

Wouldn't a subsequent pod creation using the same volume be blocked from completing ?

Reply:
@thockin - so we didn't discuss this part, but when writing down things I remembered uniqueness of statefulset pods
re: creation blocked from completing, that was initially my thought, but if you're using a storage provider that supports `ReadWriteMany` then wouldn't that complete anyway ?

I see 2 paths here:
- when enabling `safeDetach` at the `CSIDriver` level, state that `statefulset` with `PVCs` that have `ReadWriteMany` can cause a duplication, this should avoid touching the `statefulset` APIs
- use what's proposed here to enable folks opt-in per-workload

Reply:
removed the `stateful` parts as they aren't needed